edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs

```diff
@@ -162,6 +162,38 @@ public bool RemoveSubscription(string id, DeviceSubscription deviceSubscription)
             return hasChanged;
         }
 
+        public IReadOnlyCollection<DeviceSubscription> RemoveSubscriptions(string id)
+        {
+            if (!this.devices.TryGetValue(Preconditions.CheckNonWhiteSpace(id, nameof(id)), out ConnectedDevice device))
+            {
+                throw new ArgumentException($"A connection for {id} not found.");
+            }
+
+            var toRemove = new List<DeviceSubscription>();
+            device.DeviceConnection.Filter(d => d.IsActive)
```

Comment:
the update is not thread safe, it looks like it is possible to remove a subscription if in the meantime was added from a different thread. I don't think there is a lock on a higher level to make sure it's thread safe

Reply:
This will be a bit long, sorry:

let's say that there is a lock that Remove/AddSubscription and Subscriptions use. If there is a parallel call, that let's say A) thread turns on C2D and B) thread turns off, then two things can happen:

1) A thread gets the lock first, it turns on C2D. Then B turns it off
2) B thread turns off everything, then A turns C2D on.

At this level (RemoveSubscriptions), without the lock, the same two cases can happen:
1) A thread turns C2D on before B thread gets there in the loop, so at the end, everything will be turned off
2) A thread turns off C2D after B thread got there in the loop, so at the end C2D will be turned on.

The problem in both cases (with or without lock) is that the toRemove array may have an item which is at the end is not turned off. But a local locking will not solve this, because even if we ensure that at the time of exiting the function toRemove represents the current state, right after the return something can turn a subscription back.

This problem existed before RemoveSubscriptions() was added:

Let's say that a thread removed C2D using the old ConnectionManager.RemoveSubscription() method. In the code it happens in SubscriptionProcessorBase::RemoveSubscription() method. Now in ConnectionManager the subscription is turned off.
Then let's say that another thread turned this back using AddSubscription(), so now C2D is on according to ConnectionManager.
What we cannot guarantee that in SubscriptionProcessorBase, the subsequent HandleSubscription() call will be in the same order. It can happen that for the second thread it gets executed sooner, so it will subscribe at IoTHub for C2D, and then the first thread unsubscribes. At the end, ConnectionManager and IoTHub will be out of sync.

Adding RemoveSubscriptions() did not bring it a new problem just made an existing one more visible. I was not too concerned about this problem for two reasons:
- it is not too realistic having sudden subscriptions/unsubscriptions on the same topic from a single client. (although it can happen)
- RemoveSubscriptions() is added for more convinient support when a client disconnects and we want to remove all subscriptions. Because of this, in real life it is hard to receive a subscription in the middle of RemoveSubscriptions(), as the client disconnected.

It still might worth to fix the original problem, but I would create a ticket for that instead and leave this part of the code as is.

Reply:
As discussed, we should open a bug and fix it separately